### ISSUE TITLE
New Component: Table

### ISSUE THREAD

**Issue by iamabhishekmathur** [h2oai/wave#974]
### Table View

![image](https://user-images.githubusercontent.com/42759212/130116321-87591016-de6d-41df-969d-ff9a4eff5275.png)

## Features
- Columns should be sortable
- Should allow for 'badges' as part of field values
- Should have hover over effect
- Clicking on a row should be evented

Note: don't need the "Action" column equivalent

**Comment by iamabhishekmathur** [comment]
cc: @sandruparo @dulajra @mwysokin

**Comment by mtanco** [comment]
@mturoci here is what I see as the major changes from current `ui.table` object:
* Overall style / format
* Style on hover
* "badges" inside of table cells - @mturoci I don't believe there is currently a badges component, should this be a separate request?
* Hover text per row

Related UI Table Requests to consider implementing during this component redesign :
* #215 #897 for actions / commands in the last column of a table 
* #920 #646 to optionally handle table requests in the serve function (sort filter etc) 
* #602 select all for table filtering
* #428 clickable links in tables

**Comment by mturoci** [comment]
> I don't believe there is currently a badges component, should this be a separate request

I can do this as part of this issue by introducing a new `table_cell_type` (like we have icon, progress).

> Hover text per row

Not sure about this one, do you want to have a tooltip displayed when hovered over row? Can you show an example usecase maybe?

The rest of requests seems reasonable for me, thanks @mtanco!

**Comment by mtanco** [comment]
Oops, I misunderstood. Just a hover effect rather than hover text. The Table component already does this so it is just a design update.

**Comment by mturoci** [comment]
Redesign so far:
![image](https://user-images.githubusercontent.com/64769322/132531479-4f418694-f37a-4b89-a483-50b09131047b.png)
![image](https://user-images.githubusercontent.com/64769322/132531622-fb9971e7-dc66-4afa-bf9d-33c4f5e400ae.png)

A few design questions - cc @sandruparo @shanaka-rajitha @shihan007 :
* row border color is probably too low contrast in the dark theme - needs work ^^
* ![image](https://user-images.githubusercontent.com/64769322/132531893-0a94a0f5-94bb-43f6-81fb-0acff088861f.png) Remember that not all columns always have an icon. This inconsistency can cause the col alignment to data look funky - data is sometimes aligned to text, sometimes to icon. I suggest sticking with a current right alignment which makes sure the data is always aligned to column text.
* ![image](https://user-images.githubusercontent.com/64769322/132532582-b1e99f20-3724-4ef4-991c-b590c63c64ed.png) - What is the purpose of this? We currently support either normal data display (see pics above) or grouped mode (see below). Not sure where this belongs or what the usecase is.

Currently, the grouped mode looks like this:
![image](https://user-images.githubusercontent.com/64769322/132532869-9364a21e-5c9d-44d3-8b5b-d4e0a18a7f55.png)
![image](https://user-images.githubusercontent.com/64769322/132532897-9dc04029-98dd-4288-9668-f8967db13717.png)
The design however proposes header duplication which I think is unnecessary as the header is always the same for all the data.
![image](https://user-images.githubusercontent.com/64769322/132533069-65cc6956-1a66-4adb-9549-bdbcdaaa546e.png)

Design team, please let me know what you think.

@lo5 here is a proposal for `badges` table col cell API:
```ts
/** Defines cell content to be rendered instead of a simple text. */
interface TableCellType {
  /** Renders a progress arc with a percentage value in the middle. */
  progress?: ProgressTableCellType
  /** Renders an icon. */
  icon?: IconTableCellType
  /** 
   * Renders collection of chips, usually used for rendering state values.
   * Note: In case of multiple tags per row, make sure the row values are
   * separated by "," within a single cell string.
   * E.g. ui.table_row(name='...', cells=['cell1', 'BADGE1,BADGE2']).
   */
  badge?: BadgeTableCellType
}

/**
 * Create a cell type that renders a column's cells as badges instead of plain text.
 * The common usecase is to make easy distincions between a limited set of column values.
 * E.g. for state column, the corresponding badges could be: PENDING, WAITING, DONE.
*/
export interface BadgeTableCellType {
  /** Badges to be rendered. */
  badges?: Badge[]
}

/**
 * Create a badge.
 */
export interface Badge {
  /** Badge's text. */
  label: S
  /** Badge's background color. */
  color: S
}
```
The problem is that we want to support multiple badges per col. The only way to achieve this I could come up with was requiring people to serialize multiple badges within a single string. `,` is currently used as delimiter, but this can be configurable if needed ofc. Second problem is in determining which color belongs to which badge. For this I proposed specifying a list of badges that serves as a Map to get the correct badge config. Wdyt?

**Comment by shihan007** [comment]
@mturoci for table I have used naturallight color for header and border did you use the same? 
https://xd.adobe.com/view/0790f950-abbd-41fa-b372-332295fd876f-52c3/screen/4ef1b0c3-50cc-4782-9400-193c438cdc41/

**Comment by mturoci** [comment]
Yes, that's the one I used. Same as header. The contrast is too low also in the design link posted.

**Comment by shihan007** [comment]
@mturoci okay then for borders lets go with neutralTeritaryAlt. I will check on icons

**Comment by shihan007** [comment]
@mturoci  okay then for borders lets go with neutralTertiaryAlt. I will check on icons

**Comment by shihan007** [comment]
@mturoci can we change the placement of sort icon to left if we using the fluent icon? because some column header may have sort and filter dropdown too..

**Comment by mturoci** [comment]
> okay then for borders lets go with neutralTeritaryAlt.

Not much of a difference.
![image](https://user-images.githubusercontent.com/64769322/132643210-eb0d2789-772b-469f-9d1d-58e450830511.png)
I think we should probably increase the border width. I noticed that XD is somewhat not so exact on rendering the correct px sizes. They appear fine in XD, but in reality, they are much smaller.

Here is a `2px` border width with original `neutralLight` color:

![image](https://user-images.githubusercontent.com/64769322/132643746-04be2fcb-a972-4a1d-9824-1d7fa59e5f59.png)

wdyt @shihan007 ?

> can we change the placement of sort icon to left if we using the fluent icon? because some column header may have sort and filter dropdown too.

What's the reasoning? As I described in a comment, that would break alignment with data.

**Comment by shihan007** [comment]
@mturoci For some column headers sort and filter both needed in that case better to place sort icon to left. And each and every column does come with sort feature right?

**Comment by mturoci** [comment]
> For some column headers sort and filter both needed in that case better to place sort icon to left.

But why placing it to the left? What would be the benefit?

> And each and every column does come with sort feature right?

Every column can be specified as `sortable` / `filterable`, that's right.

**Comment by sandruparo** [comment]
@mturoci  I agree with @shihan007 Shihan. I think it's good to place the sorting option left align. anyways most of all the tables in the wave app store needs tables with sorting options.  So i dont think it's funky if we can align it nicely to the icon.cause icon should always be there. 

and telling about the rest of the implementation you did, - it's great Martin. great improvement. But i still see many color difference and spacing issues compared to the design. 

![image](https://user-images.githubusercontent.com/9942408/132650139-d83ff1ab-af09-4c20-a4f5-4276be5db421.png)
![image](https://user-images.githubusercontent.com/9942408/132650467-1cb4fe7f-2fb4-48cd-8f82-b0a3b1ac5864.png)
 
1. Scroller 
2. Row height 
3. Row highlight colors ( wondering whether you apply the theme Shihan has given you ) 
4. Font Colors 
5. Overall i feel we need to match everything with the colors Shihan gave you. 

Functionality  wise i feel you have achieved almost it. But i see look and feel is different still.  I just want to know is it possible to update it's theme ?
 
@mturoci @mtanco

**Comment by mturoci** [comment]
> So i dont think it's funky if we can align it nicely to the icon.cause icon should always be there.

This is not true. See the example above for example. The making `Done` column sortable would make no sense. Also the column sortability is determined based on whether it makes sense for the concrete column or not (the app dev decides).

> Scroller

This is a default browser scrollbar for overflowing content. Some browsers support styling it if you mean this, but the [browser support](https://caniuse.com/css-scrollbar) is not great at all thus going for the default seems the most reasonable in order to deliver the same experience across multiple browsers.

> Row height

This one is dynamic based on data and colum types (e.g. in this case progress table cell type makes the whole row bigger). I can set `min-height` to the one used in designs (`48px`) though, good point.

> Row highlight colors ( wondering whether you apply the theme Shihan has given you )

Yes, these are exactly the colors from the theme. Note that the theme in the screenshot is our current `neon` theme if you are wondering why the there is green instead of yellow. As we have discussed countless times, we need the designs to be working with any theme provided so concrete colors don't matter (e.g. whether the color is green or yellow in this case). The important thing is that `primary` color is used and the contrast is good enough.

**Comment by sandruparo** [comment]
1. This is not true. See the example above for example. Making Done column sortable would make no sense. Also the column sortability is determined based on whether it makes sense for the concrete column or not (the app dev decides). - 

**---- Why do we need a " Done " column ? if we need to show the user an indication the particular work in the row is  done or some states , in UX perspective we have given more suggestions for that. I dont think we need a column like that ?**

2. Also the column sortability is determined based on whether it makes sense for the concrete column or not (the app dev decides)  

----- Agree. Sure we can give a personalised configuration to the user to define those , t's setting in the fronted or backend ? 

3. This is a default browser scrollbar for overflowing content. Some browsers support styling it if you mean this, but the browser support is not great at all thus going for the default seems the most reasonable in order to deliver the same experience across multiple browsers. 

----- i would like to have a scroll defined to us, an unique one based on the design ( those requirements are not major ones,  maybe in the future ) 

4. This one is dynamic based on data and colum types (e.g. in this case progress table cell type makes the whole row bigger). I can set min-height to the one used in designs (48px) though, good point.

---- Please add a minimum height. yes.


5. Yes, these are exactly the colors from the theme. Note that the theme in the screenshot is our current neon theme if you are wondering why the there is green instead of yellow. As we have discussed countless times, we need the designs to be working with any theme provided so concrete colors don't matter (e.g. whether the color is green or yellow in this case). The important thing is that primary color is used and the contrast is good enough. 

-- GREAT !  But i would like to see how this will look when we add our YELLOW theme. Cause Right now it matters to me a lot, we as the design team want to see it's look and feel. for the countless time im requesting on that too. 🤓 🤷‍♀️🙏

**Comment by mturoci** [comment]
> ---- Why do we need a " Done " column ? if we need to show the user an indication the particular work in the row is done or some states , in UX perspective we have given more suggestions for that. I dont think we need a column like that ?

I am talking about icons columns in general. The point stands, it does not make sense to make each column sortable.

> ----- Agree. Sure we can give a personalised configuration to the user to define those , t's setting in the fronted or backend ?

Wave is backend only that allows you to draw a UI. The config is already in place. The comment was about the sorting icon placement.

> ----- i would like to have a scroll defined to us, an unique one based on the design ( those requirements are not major ones, maybe in the future )

Not sure I understood, what definitions do you expect?

See my comment about browser support (this means the scrollbar will not be the same across multiple browsers / browser versions so 2 different users might end up with differently looking UI). IMO consistency across browsers (for every user) > multiple variations for scrollbar styles or do you have something else in mind?

> -- GREAT ! But i would like to see how this will look when we add our YELLOW theme. Cause Right now it matters to me a lot, we as the design team want to see it's look and feel. for the countless time im requesting on that too. 🤓 🤷‍♀️🙏

I can send you a screenshot with the theme from designs although this will not eliminate the need for changes requested above so I don't see much value in it.

Here you go:
![image](https://user-images.githubusercontent.com/64769322/132678148-51144463-934b-4ecd-bfe8-ae597ae17516.png)

**Comment by shihan007** [comment]
@mturoci can we change hover bg to neutralLight and do the adjustment on spacing as @sandruparo mentioned.

**Comment by mturoci** [comment]
Sure, done.

**Comment by shihan007** [comment]
@mturoci yeah row height. Can't we have set a row height to a fixed value?

**Comment by mturoci** [comment]
Not really, see my comment above.

If set statically, it would look like this. We also might support longer text with wrapping in future, so static height would not work there.

Setting to `48px` looks like this (progress arc overflows)
![image](https://user-images.githubusercontent.com/64769322/132707784-8637385d-3342-46b0-b5e9-2a0fb9a0e2df.png)

**Comment by shihan007** [comment]
@mturoci table looks great now. you working separately on badges? we have some changes in badges.

**Comment by sandruparo** [comment]
@mturoci looks good now. Thanks for the efforts, i like how it looks now. Can we play with badges colors ?

**Comment by mturoci** [comment]
Badges colors are configurable - any color can be specified.

Any news about **group by** table designs I wrote about above?

**Comment by sandruparo** [comment]
Yes that also i read @mturoci  - grouping is fine but can we have nested groups within another group sets ? Like a tree hierarchy? If this component supports that and if it can has the style we gave we are fine. 

Well and how about drawing those lines ? hierarchy lines ?

**Comment by mturoci** [comment]
> grouping is fine but can we have nested groups within another group sets 

Yes, will be done as part of #961

> Well and how about drawing those lines ? hierarchy lines ?

as @lo5 suggested in the issue linked above, as of interest of time, we will postpone it and work on more impactful stuff.

> If this component supports that and if it can has the style we gave we are fine.

The style given needs more work, please see my comment.

Back to original issues that need to be addressed:
![image](https://user-images.githubusercontent.com/64769322/132818435-4aabaeb4-7577-416b-88c4-a00878a7c4dd.png)


Do you think the contrast is good enough for this theme?

Also sending a screenshot of default theme just FYI

![image](https://user-images.githubusercontent.com/64769322/132818484-904a6471-61bc-4dc1-a5a7-ee51f678f4b9.png)

**Comment by shihan007** [comment]
@mturoci yeah contrast looks okay to me text are visible over background. @sandruparo Wdyt?

**Comment by mturoci** [comment]
I meant contrast between table / row borders and background.

**Comment by sandruparo** [comment]
looks fine for me too. 


1. Yes, will be done as part of #961 

_**- sure Fine by me.**_ 


2. as @lo5 suggested in the issue linked above, as of interest of time, we will postpone it and work on more impactful stuff.

_**- sure Fine by me.**_  🥳
